File: Data Structures & Algorithms/pacific-atlantic-water-flow/submission-0.py
class Solution:
    def pacificAtlantic(self, heights: List[List[int]]) -> List[List[int]]:
        if not heights:
            return []

        rows, cols = len(heights), len(heights[0])
        pac,alt=set(),set()
        def dfs(r,c,parent,visited):
            if r<0 or c<0 or r>=rows or c>=cols:
                return
            if (r,c) in visited:
                return
            if heights[r][c]<parent:
                return
            visited.add((r,c))
            dfs(r-1,c,heights[r][c],visited)
            dfs(r+1,c,heights[r][c],visited)
            dfs(r,c-1,heights[r][c],visited)
            dfs(r,c+1,heights[r][c],visited)

                # Pacific (top row and left column)
        for c in range(cols):
            dfs(0, c, heights[0][c], pac)

        for r in range(rows):
            dfs(r, 0, heights[r][0], pac)

        for c in range(cols):
            dfs(rows-1,c,heights[rows-1][c],alt)
        for r in range(rows):
            dfs(r,cols-1,heights[r][cols-1],alt)
        result=[]
        for r in range(rows):
            for c in range(cols):
                if (r,c) in alt and (r,c) in pac:
                    result.append([r,c])
        return result

        

File: Data Structures & Algorithms/min-cost-climbing-stairs/submission-0.cpp
class Solution {
public:
    int dfs(int i, vector<int>& cost, vector<int>& dp) {
        int n = cost.size();

        if (i >= n)
            return 0;

        if (dp[i] != -1)
            return dp[i];

        dp[i] = cost[i] + min(dfs(i + 1, cost, dp),
                              dfs(i + 2, cost, dp));

        return dp[i];
    }

    int minCostClimbingStairs(vector<int>& cost) {
        int n = cost.size();
        vector<int> dp(n, -1);

        return min(dfs(0, cost, dp), dfs(1, cost, dp));
    }
};